### ISSUE TITLE
High RAM usage

### ISSUE THREAD

**Issue by pHiney** [PoESkillTree/PoESkillTree#538]
![image](https://user-images.githubusercontent.com/4302241/66865373-d5cadd00-efe2-11e9-8838-b6147dad77b2.png)

Just noting how RAM is being used

It would be nice if it could use less.

1G here, PoE has between 1.9G (at startup) to 4G or so during play. Not much left on an 8G machine.

thanks
P

**Comment by EmmittJ** [comment]
This is a limitation of the windows drawing library. We might be able find a way to reduce loading

**Comment by EmmittJ** [comment]
This is mostly a limitation of the windows drawing library. We need to load up all the textures to for the tree to render. These textures are fairly large, but there might be some optimizations we could do as I haven't really looked into the programs memory usage

**Comment by brather1ng** [comment]
I took a quick look at startup memory allocations with a profiler. My run allocated around 6.5 GB until the startup progress overlay closed. The Task Manager showed at most ~1 GB during that time, meaning most of it is garbage collected automatically.

That indicates that should your memory actually run full, the .NET GC would be able to collect much more of that memory. The actually required memory by the program is likely much lower. The excess memory is simply never collected because there's no need for it.

Looking at a heap dump I took after startup, it only shows 460 MB of 1.2 GB allocated by the process. Not sure how that is counted exactly, but I'm interpreting it such that everything not in those 460 MB is either memory that could be collected if required or is allocated by something outside our control (e.g. WPF). 290 MB of those are taken up by the calculation graph for computation (mostly because every node of the skill tree is fed into it, regardless of it being skilled or not). Another 80 MB are taken up by GameData (most of which are the stat translations because they contain a lot of duplicated information). Only then comes memory allocated by non-engine code, led by 57 MB for the SkillNodes.

At these values (460 MB non-collectable allocations controlled by us), I don't see any problem that's worth spending time on. It's probably possible to reduce that value, but I don't think we gain anything by shoving off like 100 MB.
